src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -1,1025 +1,54 @@
-                ServiceContext.ThreadPool.UnsafeRun(state => ((Http2Stream)state).CancelRequestAbortedToken(), this);
-            }
-        }
-
-        public abstract Task ProcessRequestAsync();
-
-        public void OnStarting(Func<object, Task> callback, object state)
-        {
-            lock (_onStartingSync)
-            {
-                if (HasResponseStarted)
-                {
-                    ThrowResponseAlreadyStartedException(nameof(OnStarting));
-                }
-
-                if (_onStarting == null)
-                {
-                    _onStarting = new Stack<KeyValuePair<Func<object, Task>, object>>();
-                }
-                _onStarting.Push(new KeyValuePair<Func<object, Task>, object>(callback, state));
-            }
-        }
-
-        public void OnCompleted(Func<object, Task> callback, object state)
-        {
-            lock (_onCompletedSync)
-            {
-                if (_onCompleted == null)
-                {
-                    _onCompleted = new Stack<KeyValuePair<Func<object, Task>, object>>();
-                }
-                _onCompleted.Push(new KeyValuePair<Func<object, Task>, object>(callback, state));
-            }
-        }
+        public override bool IsUpgradableRequest => false;
 
-        protected async Task FireOnStarting()
-        {
-            Stack<KeyValuePair<Func<object, Task>, object>> onStarting = null;
-            lock (_onStartingSync)
-            {
-                onStarting = _onStarting;
-                _onStarting = null;
-            }
-            if (onStarting != null)
-            {
-                try
-                {
-                    foreach (var entry in onStarting)
-                    {
-                        await entry.Key.Invoke(entry.Value);
-                    }
-                }
-                catch (Exception ex)
-                {
-                    ReportApplicationError(ex);
-                }
-            }
-        }
+        protected override string CreateRequestId()
+            => StringUtilities.ConcatAsHexSuffix(ConnectionId, ':', (uint)StreamId);
 
-        protected async Task FireOnCompleted()
+        protected override void OnReset()
         {
-            Stack<KeyValuePair<Func<object, Task>, object>> onCompleted = null;
-            lock (_onCompletedSync)
-            {
-                onCompleted = _onCompleted;
-                _onCompleted = null;
-            }
-            if (onCompleted != null)
-            {
-                foreach (var entry in onCompleted)
-                {
-                    try
-                    {
-                        await entry.Key.Invoke(entry.Value);
-                    }
-                    catch (Exception ex)
-                    {
-                        ReportApplicationError(ex);
-                    }
-                }
-            }
+            ExtraFeatureSet(typeof(IHttp2StreamIdFeature), this);
```

Comment:
I think we need to generate fields and have virtual calls to FastFeatureGet/Set instead of always creating a MaybeExtra list.